garden-backend-service/src/modal/utils.py

```diff
@@ -0,0 +1,81 @@
+from datetime import datetime
+
+from modal_proto import api_pb2
+from sqlalchemy.ext.asyncio import AsyncSession
+
+import modal
+from modal._utils.grpc_utils import retry_transient_errors
+from src.api.dependencies.database import get_db_session_maker
+from src.config import Settings
+from src.exceptions.modal import ModalException
+from src.models.modal.invocations import InvocationStatus, ModalInvocation
+from src.models.modal.modal_function import ModalFunction
+
+from .usage import estimate_usage
+
+
+async def cancel_modal_invocation(invocation: ModalInvocation, client: modal.Client):
+    request: api_pb2.FunctionCallCancelRequest = api_pb2.FunctionCallCancelRequest(
+        function_call_id=invocation.function_call_id,
+        terminate_containers=True,
+    )
+    await retry_transient_errors(client.stub.FunctionCallCancel, request)
+
+
+async def resolve_modal_invocation(
+    invocation: ModalInvocation,
+    status: InvocationStatus,
+    session: AsyncSession,
+):
+    if func := await ModalFunction.get(session, id=invocation.function_id):
+        invocation.date_resolved = datetime.now()
+        execution_time = (
+            invocation.date_resolved - invocation.date_invoked
+        ).total_seconds()
+        invocation.estimated_usage = estimate_usage(func, execution_time)
+        invocation.status = status
+        await session.commit()
+    else:
+        raise ValueError(f"No Modal Function with id: {id} found!")
+
+
+async def monitor_modal_invocation(
+    invocation: modal.functions._Invocation,
+    db_invocation: ModalInvocation,
+    client: modal.Client,
+    settings: Settings,
+):
+    session_maker = await get_db_session_maker(settings=settings)
+    async with session_maker() as session:
+        if inv := await ModalInvocation.get(session, id=db_invocation.id):
+            try:
+                # Try and get the invocation outputs
+                outputs_response = await invocation.pop_function_call_outputs(
+                    timeout=settings.MODAL_TIMEOUT_SECONDS,
```

Comment:
I don't follow what's happening here. If MODAL_TIMEOUT_SECONDS is 5 seconds by default, does the Modal function have a maximum of 5 seconds to run?

Reply:
`pop_function_call_outputs` ([src](https://github.com/modal-labs/modal-client/blob/2aedf2b1e40a61ffa09ae2392bd7c62087e498d8/modal/functions.py#L148)) will poll Modal's backend for the outputs from an invocation based on the `timout` parameter. If the timeout is reached before the invocation has completed, `pop_function_call_outputs` returns a response object with a non-zero value in `num_unfinished_inputs` [(src).](https://github.com/modal-labs/modal-client/blob/2aedf2b1e40a61ffa09ae2392bd7c62087e498d8/modal_proto/api.proto#L1442) To cancel the invocation and shutdown the containers we can send a `FuncionCallCancelRequest` [(src).](https://github.com/modal-labs/modal-client/blob/2aedf2b1e40a61ffa09ae2392bd7c62087e498d8/modal_proto/api.proto#L1255)


I bumped the default for `MODAL_TIMEOUT_SECONDS` to 300 seconds, does that seem like decent default?

Reply:
Thank you for spelling that out! I think 300 seconds is a decent place to start once the async routes are in place. For the purpose of this PR, maybe 20 or 25 seconds makes sense to come in under the gateway timeout that is currently in place on the FastAPI app (which I think is 30 seconds)